docs: add new feature explaination on guest cluster for load balancer (#1082)
harvester/docs: docs/rancher/cloud-provider.md

- **DHCP:** A DHCP server is required. The Harvester load balancer will request an IP address from the DHCP server.

>   Starting with **Rancher v2.15.1**, you can select a **VM network** from the UI when creating a `LoadBalancer` service. This tells the load balancer which network interface to use, so the VIP is bound to the correct interface instead of always defaulting to the management interface. If no VM network is selected, the behavior is the same as before.

Reviewer: need to mention currently the UI selection defaults to `kube-vip` for the selected network, if user disables `kube-vip` but uses another loadbalancer provider, then user needs to check for related solution.

Author: I'll put this on limitation section.

>   Starting with **Rancher v2.15.1**, you can select a **VM network** from the UI when creating a `LoadBalancer` service. This tells the load balancer which network interface to use, so the VIP is bound to the correct interface instead of always defaulting to the management interface. If no VM network is selected, the behavior is the same as before.

Reviewer: another two limitations:
(1) The pre-condition of LB on secondary-network is: each guest-cluster node (Harvester VM) has successfully set the second-network NIC, which has IP and route. If not, the LB might not work. and this is also the first step to troubeshooting.

(2) If user wants to change the LB network, user needs to delete and recreate it. Direct change the network annotation might have potential issues. (I did not test this yet, but suspect it has issues)

    On earlier Rancher versions (v2.12.x, v2.13.x, and v2.14.x), you can achieve the same result by adding the following annotations to the `Service` manifest:

    - `cloudprovider.harvesterhci.io/ipam: "dhcp"`

>     - `cloudprovider.harvesterhci.io/network: "default/mgmt-vlan1"`

Reviewer: need the user configure the viphost related annotation manually as well?

Author: No, kube-vip will help us do that.

>   - `cloudprovider.harvesterhci.io/network: "default/mgmt-vlan1"`

Reviewer: also mention the limitation, if user sets the annotation `cloudprovider.harvesterhci.io/network` directly, but the network is wrong/incorrect, or the network topology is `Asymmetric Network Topology`, then the LB might not get IP or LB IP is un-reachable.  There is no webhook to check for this. Suggest user to select the target network from UI.

>   ![](../../static/img/v1.9/rancher/guest-cluster-load-balancer-dhcp.png)

Reviewer 2: ```suggestion
    Starting with **Rancher v2.15.1**, you can select a VM network when creating a `LoadBalancer` service using the UI. This enables explicit binding of the virtual IP to the correct network interface. If you do not select a VM network, the load balancer uses the default interface.

    On earlier Rancher versions (v2.12.x, v2.13.x, and v2.14.x), you can achieve the same result by adding the following annotations to the `Service` manifest:

    - `cloudprovider.harvesterhci.io/ipam: "dhcp"`
    - `cloudprovider.harvesterhci.io/network: "default/mgmt-vlan1"`

    ![](../../static/img/v1.9/rancher/guest-cluster-load-balancer-dhcp.png)
```

Reviewer 2: I changed the indentation because of the nested list.

- **Pool:** An [IP pool](../networking/ippool.md) must be configured first. The Harvester load balancer controller will allocate an IP for the load balancer service following [the IP pool selection policy](../networking/ippool.md#selection-policy). Notice the difference between [Create IP Pool from Harvester UI directly](../networking/ippool.md#how-to-create) and [Create IP Pool from Rancher Manager UI](../networking/ippool.md#create-ip-pool-from-rancher-manager-ui). Refer to the [Best Practice](../networking/ippool.md#best-practice).

>     Starting with **Rancher v2.15.1**, you can select a **VM network** from the UI when creating a `LoadBalancer` service to explicitly bind the load balancer to a specific network interface. If no VM network is selected, the load balancer controller automatically determines the network.

Reviewer: If no VM network (specifically, the `cloudprovider.harvesterhci.io/network` is empty) is selected

    On older Rancher versions (v2.12.x, v2.13.x, and v2.14.x), the UI does not have this option, but you can still achieve the same result by setting the following annotations on the service:

    - `cloudprovider.harvesterhci.io/ipam: "ippool"`

>     - `cloudprovider.harvesterhci.io/network: "default/mgmt-vlan1"`

Reviewer 2: ```suggestion
- **Pool:** A pre-configured [IP pool](../networking/ippool.md) is required. The Harvester load balancer controller allocates an IP for the load balancer service according to the [IP pool selection policy](../networking/ippool.md#selection-policy). You can create IP pools using either the [Harvester UI](../networking/ippool.md#how-to-create) or the [Rancher UI](../networking/ippool.md#create-ip-pool-from-rancher-manager-ui). For more information, see [Best Practices](../networking/ippool.md#best-practice).

    Starting with **Rancher v2.15.1**, you can select a VM network when creating a `LoadBalancer` service using the UI. This enables explicit binding of the load balancer to the correct network interface. If you do not select a VM network, the load balancer controller automatically determines the network to be used.

    On earlier Rancher versions (v2.12.x, v2.13.x, and v2.14.x), you can achieve the same result by adding the following annotations to the `Service` manifest:

    - `cloudprovider.harvesterhci.io/ipam: "ippool"`
    - `cloudprovider.harvesterhci.io/network: "default/mgmt-vlan1"`
```

Reviewer 2: I changed the indentation because of the nested list.

:::

#### Asymmetric Network Topology

The network dropdown list on the UI displays only networks assigned to the _exact same interface position across all cluster nodes_.

Example:

| Network-Interface Mapping | UI Behavior | Node A | Node B | Displayed Networks |
| :--- | :--- | :--- | :--- | :--- |
| Identical mapping across all nodes | All networks are displayed | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-101` | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-101` | `mgmt` and `net-101` |
| Network in different interface positions across nodes | Network is hidden | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-101` | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-102`<br/>`enp3s0` → `net-101` | `mgmt` |
| Network absent on some nodes | Network is hidden | `enp1s0` → `mgmt` | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-101` | `mgmt` |
| Swapped interface mapping order | Only matching networks are displayed | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-101`<br/>`enp3s0` → `net-102` | `enp1s0` → `mgmt`<br/>`enp2s0` → `net-102`<br/>`enp3s0` → `net-101` | `mgmt` |

:::tip

> If a VM network is attached in a different NIC order across nodes, reconfigure the network interface order in the machine pool settings and let Rancher Manager and RKE2 re-provision the guest cluster VMs.

Reviewer 2: ```suggestion
If VM network interfaces are attached in different orders across nodes, reconfigure the network interface order in the machine pool settings to allow Rancher and RKE2 to reprovision the guest cluster virtual machines.
```

> If a VM network is attached in a different NIC order across nodes, shut down the affected virtual machines, reorder the network interfaces so the attachment order is consistent across all nodes, and restart the virtual machines.

Reviewer: On guest cluster case, it is to re-configure the related machine-pool and let Rancher Manager & RKE2 re-provision the guest cluster VMs.

https://docs.harvesterhci.io/v1.8/rancher/node/rke2-cluster#create-rke2-kubernetes-cluster

:::

### Limitations

> - **Default LB provider:** The UI defaults to `kube-vip` as the selected network for load balancers. If you disable `kube-vip` and use a different load balancer provider, refer to the documentation for your chosen provider to configure it correctly.

Reviewer 2: ```suggestion
- **Default load balancer provider**: `kube-vip` is selected by default on the UI. If you disable `kube-vip` and use an alternative provider, refer to that provider's documentation for configuration instructions.
```

> - **Pre-condition for LB on secondary network:** Each guest-cluster node (Harvester VM) must have the secondary-network NIC configured with a valid IP address and route. If this is not the case, the load balancer may not function. Verifying this NIC configuration is the first step when troubleshooting secondary-network LB issues.

Reviewer 2: ```suggestion
- **Pre-condition for secondary network load balancing**: The secondary network interface of each guest cluster node must have a valid IP address and route. Otherwise, the load balancer cannot route traffic. Verifying this interface configuration should be the first step when troubleshooting issues related to secondary network load balancers.
```

> - **Changing the LB network:** If you need to change the load balancer network, delete and recreate the load balancer service. Directly modifying the network annotation on an existing service may cause unexpected behavior and is not recommended.

Reviewer 2: ```suggestion
- **Load balancer network changes**: Delete and recreate the load balancer service if you require changes to the load balancer network. Modifying the network annotation on an existing service may cause unexpected behavior and is not supported.
```

> - **Setting the network annotation incorrectly:** If you set the annotation `cloudprovider.harvesterhci.io/network` directly and the specified network is incorrect or exhibits an [Asymmetric Network Topology](#asymmetric-network-topology), the load balancer may fail to obtain an IP address or the assigned IP may be unreachable. There is no webhook validation for this annotation. It is recommended to select the target network through the UI instead.

Reviewer 2: ```suggestion
- **Incorrect network annotation**: The load balancer may fail to obtain an IP address or become unreachable if you directly configure the `cloudprovider.harvesterhci.io/network` annotation and specify a network that is either invalid or exhibits an [asymmetric topology](#asymmetric-network-topology). Because webhook validation is not performed on this annotation, select the target network on the UI instead.
```

:::

#### Asymmetric Network Topology

> The VM network dropdown only shows networks that are attached to the **same interface position** across all nodes. This means:

Reviewer 2: ```suggestion
The network dropdown list on the UI displays only networks assigned to the _exact same interface position across all cluster nodes_.

Example:
```

- **All nodes share the same networks on the same interfaces** — all networks appear in the dropdown.

  ```
  Node A: enp1s0 → management, enp2s0 → net-101
  Node B: enp1s0 → management, enp2s0 → net-101
  ```
  Both `management` and `net-101` appear in the dropdown.

- **Same network, but different interface position across nodes** — the network does **not** appear in the dropdown.

  ```
  Node A: enp1s0 → management, enp2s0 → net-101
  Node B: enp1s0 → management, enp2s0 → net-102, enp3s0 → net-101
  ```
  `net-101` is on `enp2s0` for Node A but `enp3s0` for Node B, so it will not appear.

- **Network missing on some nodes** — the network does **not** appear in the dropdown.

  ```
  Node A: enp1s0 → management
  Node B: enp1s0 → management, enp2s0 → net-101
  ```
  `net-101` is absent from Node A, so it will not appear.

- **Same networks on all nodes, but attached in a different order** — only the management network appears in the dropdown.

  ```
  Node A: enp1s0 → management, enp2s0 → net-101, enp3s0 → net-102
  Node B: enp1s0 → management, enp2s0 → net-102, enp3s0 → net-101
  ```

>   Both nodes have all three networks, but `net-101` and `net-102` are swapped. Only `management` is on the same interface across all nodes, so only `management` appears in the dropdown.

Reviewer 2: ```suggestion
| Network-Interface Mapping | UI Behavior | Node A | Node B | Displayed Networks |
| :--- | :--- | :--- | :--- | :--- |
| Identical mapping across all nodes | All networks are displayed | `enp1s0` → `mgmt`<br>`enp2s0` → `net-101` | `enp1s0` → `mgmt`<br>`enp2s0` → `net-101` | `mgmt` and `net-101` |
| Network in different interface positions across nodes | Network is hidden | `enp1s0` → `mgmt`<br>`enp2s0` → `net-101` | `enp1s0` → `mgmt`<br>`enp2s0` → `net-102`<br>`enp3s0` → `net-101` | `mgmt` |
| Network absent on some nodes | Network is hidden | `enp1s0` → `mgmt` | `enp1s0` → `mgmt`<br>`enp2s0` → `net-101` | `mgmt` |
| Swapped interface mapping order | Only matching networks are displayed | `enp1s0` → `mgmt`<br>`enp2s0` → `net-101`<br>`enp3s0` → `net-102` | `enp1s0` → `mgmt`<br>`enp2s0` → `net-102`<br>`enp3s0` → `net-101` | `mgmt` |
```

Reviewer 2: I suggest using a table for easier scanning.

> For the case where the same VM network is attached but in a different NIC order across nodes, the workaround is to shut down the affected VMs, reorder the network interfaces so the attachment order is consistent across all nodes, and restart the VMs.

Reviewer 2: ```suggestion
:::tip

If a VM network is attached in a different NIC order across nodes, shut down the affected virtual machines, reorder the network interfaces so the attachment order is consistent across all nodes, and restart the virtual machines.

:::
```
